sympy/sets/sets.py

```diff
@@ -2027,6 +2027,174 @@ def __iter__(self):
+    >>> A = FiniteSet(1, 2, 3)
+    >>> B = Interval(0, 5)
+    >>> DisjointUnion(A, B)
+    DisjointUnion(FiniteSet(1, 2, 3), Interval(0, 5))
+    >>> DisjointUnion(A, B).rewrite(Union)
+    Union(ProductSet(FiniteSet(1, 2, 3), FiniteSet(0)), ProductSet(Interval(0, 5), FiniteSet(1)))
+    >>> C = FiniteSet(Symbol('x'), Symbol('y'), Symbol('z'))
+    >>> DisjointUnion(C, C)
+    DisjointUnion(FiniteSet(x, y, z), FiniteSet(x, y, z))
+    >>> DisjointUnion(C, C).rewrite(Union)
+    ProductSet(FiniteSet(x, y, z), FiniteSet(0, 1))
+
+    References
+    ==========
+
+    https://en.wikipedia.org/wiki/Disjoint_union
+    """
+
+    def __new__(cls, *sets):
+        dj_collection = []
+        for set_i in sets:
+            if isinstance(set_i, Set):
+                dj_collection.append(set_i)
+            else:
+                raise ValueError("Invalid input: '%s', input args \
+                    to DisjointUnion must be Sets" % set_i)
+        obj = Basic.__new__(cls, *dj_collection)
+        return obj
+
+    @property
+    def sets(self):
+        return self.args
+
+    @property
+    def is_empty(self):
+        return fuzzy_and(s.is_empty for s in self.sets)
+
+    @property
+    def is_finite_set(self):
+        all_finite = fuzzy_and(s.is_finite_set for s in self.sets)
+        return fuzzy_or([self.is_empty, all_finite])
+
+    @property
+    def is_iterable(self):
+        if self.is_empty:
+            return False
+        iter_flag = True
+        for set_i in self.sets:
+            if not set_i.is_empty:
+                iter_flag = iter_flag and set_i.is_iterable
+        return iter_flag
+
+    def _eval_rewrite_as_Union(self, *sets):
+        """
+        Rewrites the disjoint union as the union of (``set`` x {``i``})
+        where ``set`` is the element in ``sets`` at index = ``i``
+        """
+
+        dj_union = EmptySet()
+        index = 0
+        for set_i in sets:
+            if isinstance(set_i, Set):
+                cross = ProductSet(set_i, FiniteSet(index))
+                dj_union = Union(dj_union, cross)
+                index = index + 1
+            else:
+                continue
+        return dj_union
+
+    def _contains(self, element):
+        """
+        'in' operator for DisjointUnion
+
+        Examples
+        ========
+
+        >>> from sympy import Interval, DisjointUnion
+        >>> D = DisjointUnion(Interval(0, 1), Interval(0, 2))
+        >>> (0.5, 0) in D
+        True
+        >>> (0.5, 1) in D
+        True
+        >>> (1.5, 0) in D
+        False
+        >>> (1.5, 1) in D
+        True
+
+        Passes operation on to constituent sets
+        """
+        if not isinstance(element, Tuple) or len(element) != 2:
+            return False
+
+        if not element[1].is_Integer:
+            return False
+
+        if element[1] >= len(self.sets) or element[1] < 0:
+            return False
+
+        return element[0] in self.sets[element[1]]
+
+    def __iter__(self):
```

Comment:
I think that this implementation is fragile and hanging for 
```
A = Range(0, 10)
B = FiniteSet(0, 1, 2)
list(DisjointUnion(A, B))
```
I'd rather implement it like
```python3
    def __iter__(self):
        if self.is_iterable:
            from sympy.core.numbers import Integer

            iters = []
            for i, s in enumerate(self.sets):
                iters.append(iproduct(s, {Integer(i)}))

            return iter(roundrobin(*iters))
        else:
            raise TypeError("'%s' is not iterable." % self)
```

Reply:
This is indeed a much neater implementation, since it reuses a lot of existing code (such as `iproduct` and `roundrobin`). But Sir may I ask, why did you cast the output of roundrobin(*iters) into iter? As in, why not
```
return roundrobin(*iters)
```
instead of
```
return roundrobin(*iters)
```

Reply:
Also, I ran the example that you gave:
```
>>> from sympy import DisjointUnion, Range, FiniteSet
>>> A = Range(0, 10)
>>> B = FiniteSet(0, 1, 2)
>>> list(DisjointUnion(A, B))
[(0, 0), (0, 1), (1, 0), (1, 1), (2, 0), (2, 1), (3, 0), (4, 0), (5, 0), (6, 0), (7, 0), (8, 0), (9, 0)]
```
It ran perfectly on my system, without hanging or anything